schnorr_fun/src/frost.rs

```diff
@@ -0,0 +1,1202 @@
+            }
+            total_secret_share = s!(total_secret_share + secret_share);
+        }
+
+        let total_secret_share = total_secret_share.expect_nonzero(
+            "since verification shares are non-zero, the total secret share cannot be zero",
+        );
+
+        Ok((total_secret_share, KeyGen.frost_key))
+    }
+
+    /// Calls [`Frost::finish_keygen`] but immediately provides the the key as an [`XOnlyFrostKey`].
+    ///
+    /// # Return value
+    ///
+    /// Your total secret share Scalar and the XOnlyFrostKey
+    pub fn finish_keygen_to_xonly(
+        &self,
+        KeyGen: KeyGen,
+        my_index: u32,
+        secret_shares: Vec<Scalar<Secret, Zero>>,
+        proofs_of_possession: Vec<Signature>,
+    ) -> Result<(Scalar, XOnlyFrostKey), FinishKeyGenError> {
+        let (secret_share, frost_key) =
+            self.finish_keygen(KeyGen, my_index, secret_shares, proofs_of_possession)?;
+        Ok((secret_share, frost_key.into_xonly_key()))
+    }
+}
+
+/// Calculate the lagrange coefficient for participant with index x_j and other signers indexes x_ms
+fn lagrange_lambda(x_j: u32, x_ms: &[u32]) -> Scalar {
+    let x_j = Scalar::from(x_j).expect_nonzero("target xcoord can not be zero");
+    x_ms.iter()
+        .map(|x_m| Scalar::from(*x_m).expect_nonzero("index can not be zero"))
+        .fold(Scalar::one(), |acc, x_m| {
+            let denominator = s!(x_m - x_j)
+                .expect_nonzero("removed duplicate indexes")
+                .invert();
+            s!(acc * x_m * denominator)
+        })
+}
+
+/// A FROST signing session
+///
+/// Created using [`Frost::start_sign_session`].
+///
+/// [`Frost::start_sign_session`]
+#[derive(Clone, Debug, PartialEq)]
+pub struct SignSession {
+    binding_coeff: Scalar,
+    nonces_need_negation: bool,
+    agg_nonce: Point<EvenY>,
+    challenge: Scalar<Public, Zero>,
+    nonces: BTreeMap<u32, Nonce>,
+}
+
+impl<H: Digest<OutputSize = U32> + Clone, NG: NonceGen + AddTag> Frost<H, NG> {
+    /// Start a FROST signing session
+    ///
+    /// ## Return value
+    ///
+    /// A FROST signing session
+    pub fn start_sign_session(
+        &self,
+        frost_key: &XOnlyFrostKey,
+        nonces: Vec<(u32, Nonce)>,
+        message: Message,
+    ) -> SignSession {
+        let mut nonce_map: BTreeMap<_, _> =
+            nonces.into_iter().map(|(i, nonce)| (i, nonce)).collect();
+
+        let agg_nonce_points: [Point; 2] = nonce_map
+            .iter()
+            .fold([Point::zero().mark::<Jacobian>(); 2], |acc, (_, nonce)| {
+                [
+                    g!({ acc[0] } + { nonce.0[0] }),
+                    g!({ acc[1] } + { nonce.0[1] }),
+                ]
+            })
+            .iter()
+            .map(|agg_nonce| {
+                agg_nonce
+                    .normalize()
+                    .mark::<NonZero>()
+                    .expect("aggregate nonce should be non-zero")
```

Comment:
it should not be but it can be :). I'd say we handle this case the same way as in `musig`.